### PR TITLE
Add missing RBI definitions for core

### PR DESCRIPTION
**Pull request by Morriar**
### Motivation

More RBI definitions for `core/`. No signatures.

### Test plan

See included automated tests.

### CODE REVIEW

**Comment by jez** [comment]
https://go/builds/bui_HMLf3YJpzjxOoh

**Comment by jez** [comment]
The problem I'm seeing when running this on Stripe's codebase:

The `array.rbi` changes shadow the superclass methods from Enumerable, making something that was previously typed now untyped on this branch.

For example `Array#max` and `Array#min` are no longer typed.

We have a bunch of code at stripe that is doing `T.must(xs.max)`, but `T.must` requires that the wrapped expression be nilable, not untyped.

Do you mind either backing out the array.rbi changes, or auditing them to not shadow the corresponding methods in enumerable.rbi?

**Comment by Morriar** [comment]
@jez I rebased this on master, solved the conflicts in exp files and removed the overloads from Enumerable. Tests are green, let me know if another method is conflicting.

**Comment by jez** [comment]
https://go/builds/bui_HMad69njteAtPq

File: rbi/core/array.rbi [hunk 1]
```diff
@@ -483,6 +483,43 @@ class Array < Object
...
+  # Equality --- Two arrays are equal if they contain the same number of
+  # elements and if each element is equal to (according to Object#==) the
+  # corresponding element in `other_ary`.
+  #
+  # ```ruby
+  # [ "a", "c" ]    == [ "a", "c", 7 ]     #=> false
+  # [ "a", "c", 7 ] == [ "a", "c", 7 ]     #=> true
+  # [ "a", "c", 7 ] == [ "a", "d", "f" ]   #=> false
+  # ```
+  def ==(_); end
```
Comment (jez) [inline rbi/core/array.rbi:521]: Every `def ==(` here that doesn't have a type overrides the `T::Boolean` return type of `BasicObject#==`

File: rbi/core/hash.rbi [hunk 2]
```diff
@@ -1032,6 +1266,58 @@ class Hash < Object
...
+  # h3 = { "b" => 357, "d" => 400 }
+  # h1.merge!(h2, h3) {|key, v1, v2| v1 }
+  #                    #=> {"a"=>100, "b"=>200, "c"=>300, "d"=>400}
+  # h1                 #=> {"a"=>100, "b"=>200, "c"=>300, "d"=>400}
+  # ```
+  #
+  # [`Hash#update`](https://docs.ruby-lang.org/en/2.6.0/Hash.html#method-i-update)
+  # is an alias for
+  # [`Hash#merge!`](https://docs.ruby-lang.org/en/2.6.0/Hash.html#method-i-merge-21).
+  def update(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:1319]: Our reflection-generated RBI says that this is variadic, not unary:

```
❯ irb
irb(main):001:0> Hash.instance_method(:update).parameters
=> [[:rest]]
```

File: rbi/core/array.rbi (hunk 1 above)
Comment (jez) [reply rbi/core/array.rbi:521]: (there are a handful of `==` methods in this PR that need to be typed or deleted)

File: rbi/core/hash.rbi [hunk 3]
```diff
@@ -570,6 +713,29 @@ class Hash < Object
...
+  # h.has_key?("a")   #=> true
+  # h.has_key?("z")   #=> false
+  # ```
+  #
+  # Note that `include?` and `member?` do not test member equality using `==` as
+  # do other Enumerables.
+  #
+  # See also
+  # [`Enumerable#include?`](https://docs.ruby-lang.org/en/2.6.0/Enumerable.html#method-i-include-3F)
+  def include?(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:735]: This overrides and untypes `Enumerable#include?`

File: rbi/core/array.rbi [hunk 4]
```diff
@@ -1177,6 +1298,27 @@ class Array < Object
...
+  # An [`Enumerator`](https://docs.ruby-lang.org/en/2.6.0/Enumerator.html) is
+  # returned if neither a block nor argument is given.
+  #
+  # ```ruby
+  # a = [ "a", "b", "c" ]
+  # a.index("b")              #=> 1
+  # a.index("z")              #=> nil
+  # a.index {|x| x == "b"}    #=> 1
+  # ```
+  def find_index(*_); end
```
Comment (jez) [inline rbi/core/array.rbi:1320]: This overrides and untypes `Enumerable#find_index`

**Comment by Morriar** [comment]
@jez I removed the remaining conflicts and fixed the signature for `update`. Let me know if something else is blocking.

**Comment by jez** [comment]
https://go/builds/bui_HMbRTRqpm4RNA3

File: rbi/core/hash.rbi [hunk 5]
```diff
@@ -147,6 +147,73 @@ class Hash < Object
...
+  # Returns `true` if *hash* is subset of *other*.
+  #
+  # ```ruby
+  # h1 = {a:1, b:2}
+  # h2 = {a:1, b:2, c:3}
+  # h1 < h2    #=> true
+  # h2 < h1    #=> false
+  # h1 < h1    #=> false
+  # ```
+  def <(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:159]:
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def <(_); end
```

File: rbi/core/hash.rbi [hunk 6]
```diff
@@ -147,6 +147,73 @@ class Hash < Object
...
+  # ```
+  #
+  # The orders of each hashes are not compared.
+  #
+  # ```ruby
+  # h1 = { "a" => 1, "c" => 2 }
+  # h2 = { "c" => 2, "a" => 1 }
+  # h1 == h2   #=> true
+  # ```
+  def ==(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:193]:
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def ==(_); end
```

File: rbi/core/hash.rbi [hunk 7]
```diff
@@ -147,6 +147,73 @@ class Hash < Object
...
+  # Returns `true` if *other* is subset of *hash*.
+  #
+  # ```ruby
+  # h1 = {a:1, b:2}
+  # h2 = {a:1, b:2, c:3}
+  # h1 > h2    #=> false
+  # h2 > h1    #=> true
+  # h1 > h1    #=> false
+  # ```
+  def >(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:204]:
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def >(_); end
```

File: rbi/core/hash.rbi [hunk 8]
```diff
@@ -147,6 +147,73 @@ class Hash < Object
...
+  # Returns `true` if *other* is subset of *hash* or equals to *hash*.
+  #
+  # ```ruby
+  # h1 = {a:1, b:2}
+  # h2 = {a:1, b:2, c:3}
+  # h1 >= h2   #=> false
+  # h2 >= h1   #=> true
+  # h1 >= h1   #=> true
+  # ```
+  def >=(_); end
```
Comment (jez) [inline rbi/core/hash.rbi:215]:
```suggestion
  sig {params(_: T.untyped).returns(T::Boolean)}
  def >=(_); end
```

**Pull request merged**